File: codingame/code-of-the-rings/complex.hs
import System.IO
import Control.Monad
import Data.Char
import Data.List


moveLeft = '<'
moveRight = '>'
increase = '+'
decrease = '-'
spell = '.'
startLoop = '['
stopLoop = ']'


a0 = ord 'A' - 1
z0 = ord 'Z' + 1
alphabetLen = z0 - a0
zonesLen = 30

letter_distance_move :: Char -> Char -> (Int, Char)
letter_distance_move l1 l2
  | l1 == l2 = (0, increase)
  | l1 == ' ' = min (l2o - a0, increase) (z0 - l2o, decrease)
  | l2 == ' ' = min (z0 - l1o, increase) (l1o - a0, decrease)
  | l1o < l2o = min (l2o - l1o, increase) (alphabetLen - l2o + l1o, decrease)
  | otherwise = min (l1o - l2o, decrease) (alphabetLen - l1o + l2o, increase)
  where
    l1o = ord l1
    l2o = ord l2


replaceAtIndex n item ls = a ++ (item:b) where (a, (_:b)) = splitAt n ls


zone_distance_move :: Int -> Int -> (Int, Char)
zone_distance_move z1 z2
  | z1 == z2 = (0, moveRight)
  | z1 < z2 = min (z2 - z1, moveRight) (zonesLen - z2 + z1, moveLeft)
  | otherwise = min (z1 - z2, moveLeft) (zonesLen - z1 + z2, moveRight)


-- zones  pos  output
data GameState = GameState [Char] Int [Char]
instance Eq GameState where
  (GameState z1 p1 o1) == (GameState z2 p2 o2) = (z1, p1, o1) == (z2, p2, o2)
instance Ord GameState where
  (GameState z1 p1 o1) `compare` (GameState z2 p2 o2) = (length o1) `compare` (length o2)

get_output (GameState zones pos output) = output


make_move :: GameState -> Int -> Char -> GameState
make_move (GameState zones pos output) newZone newLetter = 
  GameState nZones newZone (output ++ (replicate moveZoneNum moveZoneHow) ++ (replicate moveLetterNum moveLetterHow) ++ [spell])
  where
    (moveZoneNum, moveZoneHow) = zone_distance_move pos newZone
    (moveLetterNum, moveLetterHow) = letter_distance_move (zones !! pos) newLetter
    nZones = replaceAtIndex newZone newLetter zones

splits = 1
deep = 3

generate :: GameState -> [Char] -> [GameState]
generate gs [] = [gs]
generate gs words =
  intercalate [] [generate newgs (tail words)| newgs <- take splits (sort moves)]
  where
    moves = [make_move gs i (head words) | i <- [0..zonesLen-1]]


go :: [GameState] -> String -> [GameState]
go gs [] = gs
go gs magicPhrase =
  go (take 10 (sort states)) right
  where
    states = intercalate [] [generate state left | state <- gs]
    (left, right) = splitAt deep magicPhrase

main :: IO ()
main = do
    hSetBuffering stdout NoBuffering -- DO NOT REMOVE
    magicphrase <- getLine

    let states = [(GameState (replicate 30 ' ') 0 [])]
    putStrLn (get_output $ head $ go states magicphrase)

File: codingame/code-of-the-rings/complex.py
#!/bin/python

import sys
import math

MOVE_LEFT = '<'
MOVE_RIGHT = '>'
INCREASE = '+'
DECREASE = '-'
SPELL = '.'
START_LOOP = '['
STOP_LOOP = ']'

magicPhrase = input()

ALPHABET_LEN = ord('Z') - ord('A') + 1
ZONES_LEN = 30

zones = [' ' for x in range(ZONES_LEN)]
pos = 0

output = []


def calculate_distance(pos, i):
    return min(abs(pos - i), 30 - abs(pos - i))
    

def calculate_change(a, b):
    if a == ' ' and b == ' ':
        return 0
    if a == ' ':
        return min(abs(ord(b) - ord('A') - 1), abs(ord('Z') + 1 - ord(b)))
    if b == ' ':
        return min(abs(ord(a) - ord('A') - 1), abs(ord('Z') + 1 - ord(a)))
        
    return min(abs(ord(a) - ord(b)), 30 - abs(ord(a) - ord(b)))

def move_to(new_zone):
    global pos, output
    if new_zone > pos:
        if new_zone - pos < pos + ZONES_LEN - new_zone: #go right
            output.extend([MOVE_RIGHT for x in range(new_zone - pos)])
        else:
            output.extend([MOVE_LEFT for x in range(pos + ZONES_LEN - new_zone)])
    elif new_zone < pos:
        if pos - new_zone < new_zone + ZONES_LEN - pos: #go left
            output.extend([MOVE_LEFT for x in range(pos - new_zone)])
        else:
            output.extend([MOVE_RIGHT for x in range(new_zone + ZONES_LEN - pos)])
    pos = new_zone

def change_letter(new_letter):
    current = zones[pos]
    nl, c = ord(new_letter), ord(current)
    if c == nl:
        return
    elif current == ' ':
        if nl - ord('A') + 1 < ord('Z') + 1 - nl: #better to go up
            output.extend([INCREASE for x in range(nl - ord('A') + 1)])
        else:
            output.extend([DECREASE for x in range(ord('Z') + 1 - nl)])
    elif new_letter == ' ':
        if ord('Z') + 1 - c < c - ord('A') + 1: #better to go up
            output.extend([INCREASE for x in range(ord('Z') + 1 - c)])
        else:
            output.extend([DECREASE for x in range(c - ord('A') + 1)])
    else:
        if (nl > c and nl - c < c - ord('A') + ord('Z') - nl + 1) or \
           (c - nl > nl - ord('A') + ord('Z') - c + 1): #better to go up
            output.extend([INCREASE for x in range(min(abs(nl - c), abs(c - ord('A') + ord('Z') - nl + 1)))])
        else:
            output.extend([DECREASE for x in range(min(abs(nl - c), abs(nl - ord('A') + ord('Z') - c + 1)))])
        
    zones[pos] = new_letter


for letter in magicPhrase:
    best_move_zone = None
    best_result = 1000000
    for i in range(ZONES_LEN):
        d = calculate_distance(pos, i)
        change = calculate_change(zones[i], letter)
        if best_result > d + change:
            best_result = d + change
            best_move_zone = i

    move_to(best_move_zone)
    change_letter(letter)
    output.append(SPELL)

print("".join(output))
